api/revoke.go

```diff
@@ -0,0 +1,149 @@
+package api
+
+import (
+	"net/http"
+
+	"github.com/pkg/errors"
+	"github.com/smallstep/certificates/authority"
+	"github.com/smallstep/certificates/logging"
+	"golang.org/x/crypto/ocsp"
+)
+
+/*
+// RevocationReasonCodes is a map between string reason codes
+// to integers as defined in RFC 5280
+var RevocationReasonCodes = map[string]int{
+	"unspecified":          ocsp.Unspecified,
+	"keycompromise":        ocsp.KeyCompromise,
+	"cacompromise":         ocsp.CACompromise,
+	"affiliationchanged":   ocsp.AffiliationChanged,
+	"superseded":           ocsp.Superseded,
+	"cessationofoperation": ocsp.CessationOfOperation,
+	"certificatehold":      ocsp.CertificateHold,
+	"removefromcrl":        ocsp.RemoveFromCRL,
+	"privilegewithdrawn":   ocsp.PrivilegeWithdrawn,
+	"aacompromise":         ocsp.AACompromise,
+}
+
+// ReasonStringToCode tries to convert a reason string to an integer code
+func ReasonStringToCode(reason string) (int, error) {
+	// default to 0
+	if reason == "" {
+		return 0, nil
+	}
+
+	code, found := RevocationReasonCodes[strings.ToLower(reason)]
+	if !found {
+		return 0, errors.Errorf("unrecognized revocation reason '%s'", reason)
+	}
+
+	return code, nil
+}
+*/
+
+// RevokeResponse is the response object that returns the health of the server.
+type RevokeResponse struct {
+	Status string `json:"status"`
+}
+
+// RevokeRequest is the request body for a revocation request.
+type RevokeRequest struct {
+	Serial     string `json:"serial"`
+	OTT        string `json:"ott"`
+	ReasonCode int    `json:"reasonCode"`
+	Passive    bool   `json:"passive"`
+	CRL        bool   `json:"crl"`
+	OCSP       bool   `json:"ocsp"`
+	RTS        authority.RevocationTypeSelector
```

Comment:
You should add `json:"-"` without it you can unmarshal a JSON like `{"RTS": {"CRL":true}}` into that variable.

In fact, as you are passing that to `Revoke` you can just build it later in the HTTP handler, and remove it from the struct.